timely/src/dataflow/operators/enterleave.rs

```diff
@@ -90,17 +91,22 @@ impl<G: Scope, T: Timestamp+Refines<G::Timestamp>, C: Data+Container> Enter<G, T
         let (targets, registrar) = TeeCore::<T, C>::new();
         let ingress = IngressNub {
             targets: CounterCore::new(targets),
-            phantom: ::std::marker::PhantomData,
+            phantom: PhantomData,
             activator: scope.activator_for(&scope.addr()),
             active: false,
         };
         let produced = ingress.targets.produced().clone();
-
         let input = scope.subgraph.borrow_mut().new_input(produced);
-
         let channel_id = scope.clone().new_identifier();
-        self.connect_to(input, ingress, channel_id);
-        StreamCore::new(Source::new(0, input.port), registrar, scope.clone())
+        let pusher = LogPusher::new(ingress, channel_id, scope);
```

Comment:
We could skip creating `pusher` if `scope.logging()` is None. That's not what we do elsewhere, but maybe we should?

Reply:
Is there a way to do that efficiently? The only I can come up with is using a `Box<dyn Push>`, and I'm not clear about the performance implications of that. Does it trade of the `if let Som(logger)` check with a vtable lookup on each push?

```diff
  let produced = ingress.targets.produced().clone();
  let input = scope.subgraph.borrow_mut().new_input(produced);
  let channel_id = scope.clone().new_identifier();
- let pusher = LogPusher::new(ingress, channel_id, scope);
+ let pusher: Box<dyn Push<_>> = match scope.logging() {
+     Some(logger) => Box::new(LogPusher::new(ingress, channel_id, scope.index(), logger)),
+     None => Box::new(ingress),
+ };

  self.connect_to(input, pusher, channel_id);
```

Reply:
As discussed, the trick to making this efficient is to duplicate the `connect_to` call.